CIP-bytestring-map-builtin/README.md

```diff
@@ -0,0 +1,152 @@
+    - the resulting list should be ordered (in ascending order of the keys)
+    - it is linear in the size of the `ByteStringMap`
+6. `listToBSMap :: forall v . List (Pair ByteString v) -> ByteStringMap v`
+    - it transforms an association list into a map, right-biased in the case of duplicate keys
+    - it has `O(n * logn)` time complexity, where `n` is the size of the list
+    - it has linear space complexity
+    - if the input list does not contain duplicate keys and is ordered (in ascending order of the keys), the space and time complexity will be linear
+
+## Rationale: how does this CIP achieve its goals?
+
+### Constructing `ByteStringMap`s
+
+The first concern we have to address is the reasoning behind not introducing a builtin function for constructing empty `ByteStringMap`s. Since maps are polymorphic in the value type, such a function would require receiving a type tag to know which type of concrete empty map to construct. This provides for an awkward interface to users, and it can easily be avoided by delegating the responsibility of constructing empty map constants to the compiler.
+
+`insertBSMap` and `deleteBSMap` remain fundamental operations which we must expose as builtin functions.
+
+One seemingly unnecessary addition is `unionBSMap`, as it may be implemented in Plutus Core by traversing the elements of one of the maps and calling `insertBSMap`. However, as outlined by [CPS-0013][1], `ByteStringMap` should provide an optimal implementation of the Plutus Tx `Value` type, and the operations on `Value`s depend heavily on map unions. Therefore, for performance reasons, we prefer to expose this functionality as a builtin function.
+
+The final map-constructing operation we have decided to include is `listToBSMap`. The motivation for introducing this builtin is providing quick internalisation of `Data` encoded maps. The `Data` encoding of maps is list-based, and is internalised into the builtin language by decoding the data `Map` constructor as a builtin list of pairs. For maps where the key type is `ByteString`, we now offer the optimised representation of `ByteStringMap`, and it is natural that we provide the most cost-efficient way of transforming such lists of pairs into `ByteStringMap`s.
+
+### Querying `ByteStringMap`s
+
+Map lookups are essential, so we must provide `lookupBSMap`, but there remains the question of how to provide a total function with this functionality.
+
+As mentioned earlier, we have decided to base the implementation of this builtin on an upcoming feature, because it is the least invasive way of providing a total implementation and it leverages future improvements to the builtins language.
+
+With the proposed signature of `lookupBSMap`, users will be able to provide any default value for when the lookup fails, including `error`.
+
+### Traversing `ByteStringMap`s
+
+Unfortunately, due to the limitations presented earlier we cannot provide an effective builtin fold over maps.
```

Comment:
We could do it we implemented [Allow for returning application trees from builtins a.k.a. folding builtins](https://github.com/IntersectMBO/plutus/issues/6034), which is a straightforward generalization of pattern matching builtins and shouldn't be too hard to get done (famous last words).